### PR TITLE
document validation with context.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @superyyrrzz @hellosnow @ansyral @qinezh

### CODE REVIEW

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 1]
```diff
@@ -185,6 +185,40 @@ You can also use @Microsoft.DocAsCode.Common.Logger.LogWarning(System.String,Sys
 
 The different between `ReportError` and throw `DocumentException` is throwing exception will stop the build immediately but `ReportError` won't stop build but will eventually fail the build after rules are run.
 
+### Advanced: validating tokens with file context
+
+For some cases, we need to validate some token with file context.
```
Comment (hellosnow) [inline Documentation/tutorial/validate_your_markdown_files.md]: >token [](start = 41, length = 5)

tokens?

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 2]
```diff
@@ -185,6 +185,40 @@ You can also use @Microsoft.DocAsCode.Common.Logger.LogWarning(System.String,Sys
...
+
+For example, we want each topic has one title (i.e. h1 write by markdown syntax, e.g. `# <title>`).
```
Comment (hellosnow) [inline Documentation/tutorial/validate_your_markdown_files.md]: >write [](start = 55, length = 5)

writen?

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 3]
```diff
@@ -185,6 +185,40 @@ You can also use @Microsoft.DocAsCode.Common.Logger.LogWarning(System.String,Sys
...
+        re.SetPostProcess("checkH1Count", re1 =>
+        {
+            var h1Count = (int)re.GetVariable("h1Count");
+            if (h1Count != 1)
+            {
+                 Logger.LogError($"Unexpected title count: {h1Count}.");
+            }
+        })
+    });
+```
```
Comment (chenkennt) [inline Documentation/tutorial/validate_your_markdown_files.md]: explain this code (markdown rewrite engine and post process) a little bit.

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 4]
```diff
@@ -185,6 +185,46 @@ You can also use @Microsoft.DocAsCode.Common.Logger.LogWarning(System.String,Sys
...
+
+The `FromLambda` method takes two callbacks:
```
Comment (hellosnow) [inline Documentation/tutorial/validate_your_markdown_files.md]: >FromLambda [](start = 5, length = 10)

add uid to better reference?

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 5]
```diff
@@ -185,6 +185,46 @@ You can also use @Microsoft.DocAsCode.Common.Logger.LogWarning(System.String,Sys
...
 The different between `ReportError` and throw `DocumentException` is throwing exception will stop the build immediately but `ReportError` won't stop build but will eventually fail the build after rules are run.
 
+### Advanced: validating tokens with file context
+
+For some cases, we need to validate some tokens with file context.
+
+For example, we want each topic has one title (i.e. h1 written by markdown syntax, e.g. `# <title>`).
+But you cannot count them in @Microsoft.DocAsCode.MarkdownLite.IMarkdownTokenValidator, it is shared by all files, and it will be never hit when there is no heading.
+
+For this propose, we need to create validator like following:
```
Comment (hellosnow) [inline Documentation/tutorial/validate_your_markdown_files.md]: >propose [](start = 9, length = 7)

purpose

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 6]
```diff
@@ -185,6 +185,46 @@ You can also use @Microsoft.DocAsCode.Common.Logger.LogWarning(System.String,Sys
...
+                 Logger.LogError($"Unexpected title count: {h1Count}.");
+            }
+        })
+    });
+```
+
+The `FromLambda` method takes two callbacks:
+* The first will be invoked on @Microsoft.DocAsCode.MarkdownLite.MarkdownHeadingBlockToken matched in all files.
+  And the static property @Microsoft.DocAsCode.MarkdownLite.MarkdownTokenValidatorContext.CurrentRewriteEngine will provide current context object.
+* The second will be invoked on starting a new file.
```
Comment (chenkennt) [inline Documentation/tutorial/validate_your_markdown_files.md]: > a new file [](start = 40, length = 11)

will variable and postprocess be cleaned up for a new file?

**Comment by hellosnow** [comment]
:shipit:

**Comment by chenkennt** [comment]
@hexiaokai

File: Documentation/tutorial/validate_your_markdown_files.md [hunk 7]
```diff
@@ -185,6 +185,46 @@ You can also use @Microsoft.DocAsCode.Common.Logger.LogWarning(System.String,Sys
 
 The different between `ReportError` and throw `DocumentException` is throwing exception will stop the build immediately but `ReportError` won't stop build but will eventually fail the build after rules are run.
 
+### Advanced: validating tokens with file context
+
+For some cases, we need to validate some tokens with file context.
+
+For example, we want each topic has one title (i.e. h1 written by markdown syntax, e.g. `# <title>`).
+But you cannot count them in @Microsoft.DocAsCode.MarkdownLite.IMarkdownTokenValidator, it is shared by all files, and it will be never hit when there is no heading.
```
Comment (hexiaokai) [inline Documentation/tutorial/validate_your_markdown_files.md]: never be hit

File: Documentation/tutorial/validate_your_markdown_files.md (hunk 6 above)
Comment (vwxyzh) [inline Documentation/tutorial/validate_your_markdown_files.md]: yes, by GC.

---
In reply to: [74566787](https://github.com/dotnet/docfx/pull/603#discussion_r74566787) [](ancestors = 74566787)

File: Documentation/tutorial/validate_your_markdown_files.md (hunk 4 above)
Comment (vwxyzh) [inline Documentation/tutorial/validate_your_markdown_files.md]: the uid is too long ...

---
In reply to: [74565363](https://github.com/dotnet/docfx/pull/603#discussion_r74565363) [](ancestors = 74565363)

**Pull request merged**